### PR TITLE
Integrate PublicKey and SecretKey from schnorr

### LINKED ISSUE #30: Integrate PublicKey and SecretKey from Schnorr
Dusk PKI should be the central place that handles Public and Secret Keys; currently the structure are split among this repo and schnorr repo.

The Secret / Public Keys of schnorr should be moved here, and schnorr should relies on dusk-pki to handle the keys – except for `PublicKeyPair`.

### PR DESCRIPTION
**Pull request by ZER0**
- Add `PublicKey`, `PublicKeyPair` and `SecretKey` (removed from `schnorr`)
- Add `dusk_bytes::Serializable` trait to structure
- Remove manual implementation of `to_bytes` and `from_bytes`
- Remove `Error` enum
- Remove `decode` function
- Bump `dusk-jubjub` to `v0.8`
- Bump `poseidon252` to `v0.16.0`
- Bump `canonical` to `v0.5`
- Bump `canonical_derive` `v0.5`
- Update CHANGELOG to ISO 8601

Resolves: #30
See also: dusk-network/schnorr#8, dusk-network/schnorr#21

### CODE REVIEW

File: .github/workflows/dusk_ci.yml [hunk 1]
```diff
@@ -65,36 +65,6 @@ jobs:
...
-      - uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: nightly
-          override: true
-      - uses: actions-rs/cargo@v1
-        with:
-          command: test
-          args: --release --no-default-features --features canon
-
```
Comment (vlopes11) [inline .github/workflows/dusk_ci.yml:97]: Why was the `no-std` tests removed?

File: Cargo.toml [hunk 2]
```diff
@@ -1,35 +1,25 @@
...
-hex = {version = "^0.4", default-features = false}
-subtle = {version = "^2.2.1", default-features = false}
-canonical = {version = "0.4", optional = true}
-canonical_derive = {version = "0.4", optional = true}
-rand = {version = "0.7", optional = true}
-sha2 = {version = "0.8", optional = true}
+rand_core = "0.5"
+dusk-jubjub = "0.8"
+poseidon252 = {git = "https://github.com/dusk-network/Poseidon252", tag = "v0.16.0", default-features = false}
+subtle = "^2.2.1"
```
Comment (vlopes11) [inline Cargo.toml:11]: Shouldn't we completely remove subtle from our stack?

File: src/lib.rs [hunk 3]
```diff
@@ -13,27 +13,25 @@
...
-pub use spend::public::PublicKey as PublicSpendKey;
+pub use keys::spend::public::PublicSpendKey;
 /// Secret Spend Key
-pub use spend::secret::SecretKey as SecretSpendKey;
+pub use keys::spend::secret::SecretSpendKey;
 /// Stealth Address
-pub use spend::stealth::{Ownable, StealthAddress};
+pub use keys::spend::stealth::{Ownable, StealthAddress};
 /// ViewKey
 pub use view::ViewKey;
```
Comment (vlopes11) [inline src/lib.rs:31]: Isn't the view key just a duplicate of the `SecretKey`? In practice, this struct is duplicated

File: src/keys/secret.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+use rand_core::{CryptoRng, RngCore};
+
+#[cfg(feature = "canon")]
+use canonical::Canon;
+#[cfg(feature = "canon")]
+use canonical_derive::Canon;
+
+#[allow(non_snake_case)]
+#[cfg_attr(feature = "canon", derive(Canon))]
+#[derive(Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, HexDebug)]
```
Comment (vlopes11) [inline src/keys/secret.rs]: We should not have `Default` implementations for keys. This is misleading

File: src/permutation.rs [hunk 5]
```diff
@@ -10,7 +10,7 @@ use poseidon252::sponge;
 
 /// Hashes a JubJub's ExtendedPoint into a JubJub's Scalar
 pub fn hash(p: &JubJubExtended) -> JubJubScalar {
-    let h = sponge::hash(&p.to_hash_inputs());
+    let h = sponge::hash(&p.to_hash_inputs()[..]);
```
Comment (vlopes11) [inline src/permutation.rs]: Why is this change necessary?

File: src/keys/spend/public.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+use canonical::Canon;
+#[cfg(feature = "canon")]
+use canonical_derive::Canon;
+
+use dusk_bytes::{DeserializableSlice, Error, HexDebug, Serializable};
+use dusk_jubjub::GENERATOR_EXTENDED;
+use subtle::{Choice, ConstantTimeEq};
+
+/// Public pair of `a·G` and `b·G` defining a [`PublicSpendKey`]
+#[derive(HexDebug, Clone, Copy)]
```
Comment (vlopes11) [inline src/keys/spend/public.rs:24]: We should implement `PartialEq` and `Eq` since there is no restriction for that and it will help to evaluate the correct and expected keys

File: src/keys/spend/public.rs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+#[cfg_attr(feature = "canon", derive(Canon))]
+pub struct PublicSpendKey {
+    A: JubJubExtended,
+    B: JubJubExtended,
+}
+
+impl PublicSpendKey {
+    /// This method is used to construct a new `PublicSpendKey` from the given
+    /// public pair of `a·G` and `b·G`
+    pub fn new(A: JubJubExtended, B: JubJubExtended) -> Self {
```
Comment (vlopes11) [inline src/keys/spend/public.rs:34]: A public key should never be created out of two isolated points. It should always be derived from a secret key, or from some form of deserialization. Hence, this method should not be public

File: src/keys/spend/public.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        let rA = G * rA;
+
+        let pk_r = rA + self.B;
+        let pk_r = PublicKey(pk_r);
+
+        StealthAddress { R, pk_r }
+    }
+}
+
+impl ConstantTimeEq for PublicSpendKey {
```
Comment (vlopes11) [inline src/keys/spend/public.rs:64]: Why are subtle implementations desirable?

File: src/keys/spend/public.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+impl PartialEq for PublicSpendKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(&other).into()
+    }
+}
+
+impl Eq for PublicSpendKey {}
+
+impl Default for PublicSpendKey {
```
Comment (vlopes11) [inline src/keys/spend/public.rs]: Default implementation for public keys is never gonna be used and is misleading

File: src/keys/spend/stealth.rs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+use canonical_derive::Canon;
+use dusk_bytes::{DeserializableSlice, Error, HexDebug, Serializable};
+
+use subtle::{Choice, ConstantTimeEq};
+
+/// To obfuscate the identity of the participants, we utilizes a Stealth Address
+/// system.
+/// A `StealthAddress` is composed by a one-time public key (`pk_r`, the actual
+// address) and a random point `R`.
+#[derive(HexDebug, Clone, Copy)]
```
Comment (vlopes11) [inline src/keys/spend/stealth.rs:20]: Same as above; implement `PartialEq` and `Eq`

File: src/keys/spend/stealth.rs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        &self.pk_r
+    }
+
+    /// Gets the underline `JubJubExtended` point of `pk_r`
+    pub fn address(&self) -> &JubJubExtended {
+        &self.pk_r.as_ref()
+    }
+}
+
+impl ConstantTimeEq for StealthAddress {
```
Comment (vlopes11) [inline src/keys/spend/stealth.rs:51]: Same as above; why subtle?

File: src/keys/public.rs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+impl From<&SecretKey> for PublicKey {
+    fn from(sk: &SecretKey) -> Self {
+        let public_key = GENERATOR_EXTENDED * sk.0;
+
+        PublicKey(public_key)
+    }
+}
+
+impl From<JubJubExtended> for PublicKey {
```
Comment (vlopes11) [inline src/keys/public.rs:30]: Same as above; A public key should never be created directly from a point

**Review by vlopes11 — CHANGES_REQUESTED** [review]
Few nits

File: src/lib.rs (hunk 3 above)
Comment (ZER0) [reply src/lib.rs:31]: No, it's not. They have different fields.

File: Cargo.toml (hunk 2 above)
Comment (ZER0) [reply Cargo.toml:11]: As I mentioned in the rocket chat:
> There are few things to discuss over the review, I'm not sure if we still want to have subtle or not and for which types.

So it's a matter if it's important or not to have constant time comparison for our keys.

We can maybe take a decision during the coming stand down.

File: .github/workflows/dusk_ci.yml (hunk 1 above)
Comment (ZER0) [reply .github/workflows/dusk_ci.yml:97]: Because `dusk-pki` is no-std by default. There is no `std` version. So the only additional test are the one with `canon` feature set on.

Comment (vlopes11) [reply .github/workflows/dusk_ci.yml:97]: In that case, we should remove the feature `std` because it is misleading

File: src/permutation.rs (hunk 5 above)
Comment (ZER0) [reply src/permutation.rs]: As mentioned in our chat my compiler wasn't updated to the version where this is no longer necessary. I rollback the change and also updated where it was used elsewhere.

File: src/keys/spend/public.rs (hunk 8 above)
Comment (ZER0) [reply src/keys/spend/public.rs:64]: As discussed we're keeping it only for comparison, however we'll deal with general constant time operation later on on a wider scale.

File: src/keys/spend/public.rs (hunk 9 above)
Comment (ZER0) [reply src/keys/spend/public.rs]: Removed all the default impl for all keys.

File: src/keys/spend/stealth.rs (hunk 10 above)
Comment (ZER0) [reply src/keys/spend/stealth.rs:20]: Is implemented manually due the constant time comparison.

**Review by vlopes11 — APPROVED** [review]
LGTM!
